ext/mixed/js/display.js

```diff
@@ -797,11 +797,12 @@ class Display {
                 this.audioPlaying = null;
             }
 
-            let audio, source, info;
+            let audio, info;
             try {
                 const {sources, textToSpeechVoice, customSourceUrl} = this.options.audio;
-                ({audio, source} = await this.audioSystem.getDefinitionAudio(expression, sources, {textToSpeechVoice, customSourceUrl}));
-                info = `From source ${1 + sources.indexOf(source)}: ${source}`;
+                let index;
+                ({audio, index} = await this.audioSystem.getDefinitionAudio(expression, sources, {textToSpeechVoice, customSourceUrl}));
+                info = `From source ${1 + index}: ${sources[index]}`;
```

Comment:
It would be safer in general to avoid passing mutable object indices in async code. This should be fine because `sources` is destructured from options (and updated by replacing the entire object), but this could break if the last line was

```javascript
info = `From source ${1 + index}: ${this.options.audio.sources[index]}`;
```
There's a similar issue with profile index in https://github.com/FooSoft/yomichan/pull/389#discussion_r390674046

Reply:
The reason I changed it here was to avoid the now redundant `indexOf` call. For another point of reference, the backend database also returns indices corresponding to a function argument. These types of functions have a general assumption that the data being passed in is not going to be mutated during execution, and the resulting index is only intended to be consistent with the input value it received at the instance the function is invoked.

Technically, even though you mention that it "should be fine" since `sources` is destructured from `options`, it could still "break" if something like this `this.options.audio.sources.length = 0;` is called from elsewhere, since those both reference the same array. However, at least in this situation, the worst theoretical case is just something like `info = 'From source 1: undefined'`.

In general, this isn't something that can be guaranteed in all cases without frequently doing something like a deepCopy on arguments, and that comes with a non-trivial amount of overhead. Instead, like most APIs in the project, functions are relying on the developers to use them correctly, at least to a certain extent. Technically it's possible to do functional programming, but it's probably best to not even think about that :^)

Reply:
This was something fresh in mind from doing changes to options code. I agree that it would be overkill to protect against typos like `this.options.audio.sources.length = 0;` by copying the object every time. Safeguards like this should be opt out rather than opt in to work properly (copy on write and explicit reference). Functional programming can give useful tools and patterns but you have mutate the state to do anything useful :smile: